stdlib/public/runtime/ProtocolConformance.cpp

```diff
@@ -586,40 +494,31 @@ swift_conformsToSwiftProtocolImpl(const Metadata * const type,
       if (descriptor.getProtocol() != protocol)
         continue;
 
-      // If there's a matching type, record the positive result.
+      // If there's a matching type, record the positive result and return it.
+      // The matching type is exact, so they can't go stale, and we should
+      // always cache them.
       ConformanceCandidate candidate(descriptor);
-      if (candidate.getMatchingType(type)) {
-        const Metadata *matchingType = candidate.getConformingTypeAsMetadata();
-        if (!matchingType)
-          matchingType = type;
-
-        C.cacheSuccess(matchingType, protocol, &descriptor);
+      if (auto *matchingType = candidate.getMatchingType(type)) {
+        auto witness = descriptor.getWitnessTable(matchingType);
+        C.cacheSuccess(matchingType, protocol, witness, /*always cache*/ 0);
       }
     }
   }
-  
-  // Conformance scan is complete.
-
-  // Search the cache once more, and this time update the cache if necessary.
-  FoundConformance = searchInConformanceCache(type, protocol);
-  if (FoundConformance.isAuthoritative) {
-    return FoundConformance.description;
-  } else {
-    C.cacheFailure(type, protocol, snapshot.count());
-    return nullptr;
-  }
-}
 
-static const WitnessTable *
-swift_conformsToProtocolImpl(const Metadata * const type,
-                             const ProtocolDescriptor *protocol) {
-  auto description =
-    swift_conformsToSwiftProtocol(type, protocol, StringRef());
-  if (!description)
-    return nullptr;
+  // Try the search again to look for the most specific cached conformance.
+  found = searchInConformanceCache(type, protocol);
 
-  return description->getWitnessTable(
-      findConformingSuperclass(type, description));
+  // If it's not authoritative, then add an authoritative entry for this type.
+  if (!found.first) {
+    if (found.second) {
+      // This entry could be made obsolete by a more specific conformance in
+      // a new section, so only cache it if it's up to date.
+      C.cacheSuccess(type, protocol, found.second, snapshot.count());
+    } else {
+      C.cacheFailure(type, protocol, snapshot.count());
```

Comment:
See above, maybe this could just be `C.cacheSuccess(type, protocol, nullptr, snapshot.count())`??

Reply:
Good call. And since the choice of which to call is based on `found.second`, the whole thing can collapse into a single call. I'll rename it to just `cacheResult` and simplify.